components/playlist/playlist_service.cc

```diff
@@ -0,0 +1,431 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/playlist/playlist_service.h"
+
+#include <algorithm>
+#include <utility>
+
+#include "base/bind.h"
+#include "base/containers/flat_set.h"
+#include "base/files/file_enumerator.h"
+#include "base/files/file_util.h"
+#include "base/json/json_reader.h"
+#include "base/json/json_writer.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/task/post_task.h"
+#include "base/task_runner_util.h"
+#include "base/token.h"
+#include "brave/components/playlist/playlist_constants.h"
+#include "brave/components/playlist/playlist_data_source.h"
+#include "brave/components/playlist/playlist_service_helper.h"
+#include "brave/components/playlist/playlist_service_observer.h"
+#include "brave/components/playlist/playlist_types.h"
+#include "brave/components/playlist/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "services/preferences/public/cpp/dictionary_value_update.h"
+#include "services/preferences/public/cpp/scoped_pref_update.h"
+
+namespace playlist {
+namespace {
+
+const base::FilePath::StringType kBaseDirName(FILE_PATH_LITERAL("playlist"));
+
+const base::FilePath::StringType kThumbnailFileName(
+    FILE_PATH_LITERAL("thumbnail"));
+
+void DeleteDir(const base::FilePath& path) {
+  base::DeletePathRecursively(path);
+}
+
+PlaylistInfo CreatePlaylistInfo(const CreatePlaylistParams& params) {
+  PlaylistInfo p;
+  p.id = base::Token::CreateRandom().ToString();
+  p.playlist_name = params.playlist_name;
+  p.create_params = params;
+  return p;
+}
+
+std::vector<base::FilePath> GetOrphanedPaths(
+    const base::FilePath& base_dir,
+    const base::flat_set<std::string>& ids) {
+  std::vector<base::FilePath> orphaned_paths;
+  base::FileEnumerator dirs(base_dir, false, base::FileEnumerator::DIRECTORIES);
+  for (base::FilePath name = dirs.Next(); !name.empty(); name = dirs.Next()) {
+    if (ids.find(name.BaseName().AsUTF8Unsafe()) == ids.end())
+      orphaned_paths.push_back(name);
+  }
+  return orphaned_paths;
+}
+
+bool DoGenerateHTMLFileOnTaskRunner(const base::FilePath& html_file_path) {
+  constexpr char kHTMLTemplate[] =
+      "<video id='v' controls autoplay "
+      "onplay='a=document.getElementById(\"a\");a.currentTime=this.currentTime;"
+      "a.play();' onpause='a=document.getElementById(\"a\");a.pause()'><source "
+      "src='video_file.mp4' type='video/mp4' /></video> <video id='a' autoplay "
+      "style='display:none'><source src='audio_file.m4a' type='audio/mp4' "
+      "/></video>";
+
+  base::DeleteFile(html_file_path);
+  return base::WriteFile(html_file_path, kHTMLTemplate);
+}
+
+}  // namespace
+
+PlaylistService::PlaylistService(content::BrowserContext* context,
```

Comment:
we should really avoid tying these things to BrowserContext because that complicates things if we need to share with ios. Passing in BrowserContext shouldn't be a way to fetch whatever else you need inside the service. If you need UserPrefs then pass in UserPrefs. The data source should be added in the ui. If there is something that absolutely has to use something from `content` then split it out into `core` and `content` per https://www.chromium.org/developers/design-documents/layered-components-design